### PR TITLE
Safe Connections Report

### LINKED ISSUE #2124: Create safe connection report

### PR DESCRIPTION
**Pull request by madelondohmen**
### Changes
This PR adds a Safe Connections Report, to show for each IP address if all the ciphers are safe. For each IP address, it checks if any of the following finding_types were found:

* KAT-RECOMMENDATION-BAD-CIPHER
* KAT-MEDIUM-BAD-CIPHER
* KAT-CRITICAL-BAD-CIPHER

If one of those have been found, it will be visible in the rapport by showing that not all servers are compliant. Besided that, it also gives a more detailed description in a table below.

### Issue link
Closes #2124 

---

### Code Checklist
- [x] All the commits in this PR are properly PGP-signed and verified;
- [x] This PR only contains functionality relevant to the issue; tickets have been created for newly discovered issues.
- [ ] I have written unit tests for the changes or fixes I made.
- [ ] For any non-trivial functionality, I have added integration and/or end-to-end tests.
- [x] I have performed a self-review of my code and refactored it to the best of my abilities.

### Communication
- [x] I have informed others of any required `.env` changes files if required and changed the `.env-dist` accordingly.
- [x] I have made corresponding changes to the documentation, if necessary.

---
## Checklist for code reviewers:
Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_code.md) into your comment.

---
## Checklist for QA:
Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_qa.md) into your comment.

### CODE REVIEW

File: rocky/reports/report_types/safe_connections_report/report.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+from django.utils.translation import gettext_lazy as _
+
+from octopoes.models import Reference
+from octopoes.models.ooi.dns.zone import Hostname
+from octopoes.models.ooi.network import IPAddressV4, IPAddressV6
+from reports.report_types.definitions import Report
+
+CIPHER_FINDINGS = ["KAT-RECOMMENDATION-BAD-CIPHER", "KAT-MEDIUM-BAD-CIPHER", "KAT-CRITICAL-BAD-CIPHER"]
+
```
Comment (underdarknl) [inline rocky/reports/report_types/safe_connections_report/report.py:12]: For hostnames, should we also check for missing or expired certs?
KAT-NO-CERTIFICATE and KAT-CERTIFICATE-EXPIRED
possibly also KAT-CERTIFICATE-EXPIRING-SOON for soon to be expiring certs.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -56,7 +56,7 @@ repos:
   - id: ruff
 
 - repo: https://github.com/psf/black
-  rev: "23.11.0"
+  rev: "23.10.0"
```
Comment (underdarknl) [inline .pre-commit-config.yaml:59]: possibly overlooked this, im not sure you meant to downgrade Black here?

Comment (madelondohmen) [reply .pre-commit-config.yaml:59]: You're right.. I overlooked it, it shoudn't be downgraded

File: rocky/reports/report_types/safe_connections_report/report.py (hunk 1 above)
Comment (noamblitz) [reply rocky/reports/report_types/safe_connections_report/report.py:12]: That's only for web

Comment (underdarknl) [reply rocky/reports/report_types/safe_connections_report/report.py:12]: Certificates can be on any service that talks TLS (including mailservers, webservers, database servers). I think we can collect these findings on this subreport  and see where they fit in the aggregate report.

Comment (noamblitz) [reply rocky/reports/report_types/safe_connections_report/report.py:12]: In KAT certificates can only exist on websites.

Comment (underdarknl) [reply rocky/reports/report_types/safe_connections_report/report.py:12]: For now. Yes, and thats an oversight in itself. Im okay with moving this to another sprint to keep this one clean. However, lets make sure at least the cert issues show up in the aggregate report somewhere.

Comment (noamblitz) [reply rocky/reports/report_types/safe_connections_report/report.py:12]: https://github.com/minvws/nl-kat-coordination/issues/2081

Yes they will

**Review by ammar92 — APPROVED** [review]
Looks good to me; I have no remarks 👍

**Comment by noamblitz** [comment]
Works perfectly for me! Nice work!

**Comment by stephanie0x00** [comment]
## Checklist for QA:
- [x] I have checked out this branch, and successfully ran a fresh `make reset`.
- [x] I confirmed that there are no unintended functional regressions in this branch:
    - [x] I have managed to pass the onboarding flow
    - [x] Objects and Findings are created properly
    - [] Tasks are created and completed properly
- [] I confirmed that the PR's advertised `feature` or `hotfix` works as intended.

### What works:
Onboarding flow works as expected. Boefjes generate findings and tasks are scheduled. 

### What doesn't work:
* Bad ciphers are still shown as compliant in the report. This occurs for both medium and critical findings:

image

image


### Bug or feature?:
* When marking all objects as L2 and generating the report, the workflow says that the 'TestSSL' boefje is required. When you enable this boefje no findings are generated for the selected hosts, as this boefje requires IPServices, which it doesn't generate. These objects are generated by e.g. nmap, thus the requirements should also include nmap, and other dependencies that are required before nmap can run.

File: rocky/reports/report_types/safe_connections_report/report.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+class SafeConnectionsReport(Report):
+    id = "safe-connections-report"
+    name = _("Safe Connections Report")
+    description: str = _("Shows whether the IPService contains safe ciphers.")
+    plugins = {"required": ["dns-records", "testssl-sh-ciphers", "nmap-tcp"], "optional": []}
```
Comment (underdarknl) [inline rocky/reports/report_types/safe_connections_report/report.py:18]: nmap-tcp was renamed back to nmap again due to backwards compatibility.

**Comment by stephanie0x00** [comment]
Cannot reproduce the "Bad ciphers are still shown as compliant in the report." anymore, so it's ready for merge.

**Review by dekkers — APPROVED** [review]

**Review by underdarknl — APPROVED** [review]

**Pull request merged**